File: src/main/java/eu/dragoncoding/dragonbot/commands/Clear.java
package eu.dragoncoding.dragonbot.commands;

import eu.dragoncoding.dragonbot.structures.Command;
import eu.dragoncoding.dragonbot.utils.ChatUtils;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageChannel;

import java.util.ArrayList;
import java.util.List;

public class Clear implements Command {

    @Override
    public void performCommand(Message message, int subString) {
        int errorCode,
            amount = 0;

        //noinspection ConstantConditions
        if(message.getMember().hasPermission(message.getTextChannel(), Permission.MESSAGE_MANAGE)) {
            String[] args = message.getContentDisplay().substring(subString).split(" ");

            if(args.length > 1 ) {
                try {
                    amount = Integer.parseInt(args[1]);

                    message.getChannel().purgeMessages(get(message.getChannel(), amount));
                    errorCode = 0;

                } catch (NumberFormatException e) {
                    errorCode = 3;
                }
            } else {
                errorCode = 2;
            }
        } else {
            errorCode = 1;
        }

        ArrayList<String> responseMessages = new ArrayList<>();
        responseMessages.add("Purged " + amount + (amount == 1 ? " message!" : " messages!"));
        responseMessages.add("Missing Permission! (Manage Messages)");
        responseMessages.add("Invalid argument length. See '-help clear' for more information.");
        responseMessages.add("The amount of messages to delete is invalid. Pleas only use numbers.");

        ChatUtils.sendResponseToDiscord(errorCode, message.getTextChannel(), responseMessages);
    }

    public List<Message> get(MessageChannel channel, int amount) {
        List<Message> messages = new ArrayList<>();
        int i = amount + 1;

        for(Message message : channel.getIterableHistory().cache(false)) {

            if(!message.isPinned()) {
                messages.add(message);
            }
            --i;
            if(i <= 0) break;
        }

        return messages;
    }

}

File: src/main/java/eu/dragoncoding/dragonbot/commands/utilities/Clear.kt
package eu.dragoncoding.dragonbot.commands.utilities

import eu.dragoncoding.dragonbot.structures.Command
import eu.dragoncoding.dragonbot.utils.ChatUtils.sendResponseToDiscord
import eu.dragoncoding.dragonbot.utils.SettingUtils
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.MessageChannel
import net.dv8tion.jda.api.entities.TextChannel
import java.util.*

class Clear: Command {

    override val cmdLength: Int = "clear".length + 1
    override var errorCode: Int = 0

    var amount = 0

    override fun performCommand(message: Message, subString: Int) {
        if (message.member!!.hasPermission(message.textChannel, Permission.MESSAGE_MANAGE)) {
            val args = getArgs(message, subString + cmdLength)
            if (args.isNotEmpty()) {
                try {
                    amount = if (args.isNotEmpty()) { args[0].toInt() } else { 1 }
                    message.channel.purgeMessages(getMessagesToPurge(message.channel, amount + 1, SettingUtils.checkDeleteCmd(message.guild.idLong)))
                } catch (e: NumberFormatException) {
                    errorCode = 2
                }
            }
        } else {
            errorCode = 1
        }

    }

    override fun response(channel: TextChannel) {
        val responseMessages = ArrayList<String>()

        responseMessages.add("Purged " + amount + if (amount == 1) " message!" else " messages!")
        responseMessages.add("Missing Permission! (Manage Messages)")
        responseMessages.add("The amount of messages to delete is invalid. Please only use numbers.")

        sendResponseToDiscord(errorCode, channel, responseMessages)
    }

    fun getMessagesToPurge(channel: MessageChannel, amount: Int, delCmd: Boolean): List<Message> {
        val messages: MutableList<Message> = ArrayList()
        var i = amount
        for (message in channel.iterableHistory.cache(false)) {
            if (!delCmd && i == (amount + 1)) {
                i--
                continue
            }

            if (!message.isPinned) {
                messages.add(message)
            }
            --i
            if (i <= 0) break
        }
        return messages
    }
}